File: Java/CountingAnagrams.java
import java.util.*;

public class CountingAnagrams {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> list = new ArrayList<>();
        if (s == null || s.length() == 0 || p == null || p.length() == 0) return list;
        int[] hash = new int[256]; //character hash

        for (char c : p.toCharArray()) {
            hash[c]++;
        }
        int left = 0, right = 0, count = p.length();

        while (right < s.length()) {
            if (hash[s.charAt(right)] >= 1) {
                count--;
            }
            hash[s.charAt(right)]--;
            right++;

            if (count == 0) {
                list.add(left);
            }
            if (right - left == p.length() ) {
                if (hash[s.charAt(left)] >= 0) {
                    count++;
                }
                hash[s.charAt(left)]++;
                left++;
            }
        }
        return list;
    }

    public static void main(String[] args) {
        String s = "cbaebabacd";
        String p = "abc";
        CountingAnagrams ca = new CountingAnagrams();
        List<Integer> list = ca.findAnagrams(s, p);
        StringBuilder sb = new StringBuilder();
        for (Integer i : list)
        {
            sb.append(i);
            sb.append(" ");
        }
        System.out.println(sb.toString());

        String s1 = "abab";
        String p1 = "ab";
        List<Integer> list1 = ca.findAnagrams(s1, p1);
        StringBuilder sb1 = new StringBuilder();
        for (Integer i : list1)
        {
            sb1.append(i);
            sb1.append(" ");
        }
        System.out.println(sb1.toString());
    }
}

File: Python/CountingAnagrams.py
class Solution(object):
	def findAnagrams(self, s, p):
		result = []
		hash = [0] * 128
		if not s or not p or len(s) == 0 or len(p) == 0:
			return result
		for c in p:
			hash[ord(c)] += 1
		l, r, count = 0, 0, len(p)
		while r < len(s):
			if hash[ord(s[r])] >= 1:
				count -= 1
			hash[ord(s[r])] -= 1
			r += 1

			if count == 0:
				result.append(l)

			if r - l == len(p):
				if hash[ord(s[l])] >= 0:
					count += 1
			hash[ord(s[l])] += 1
			l += 1
		return result

if __name__ == '__main__':
	s1, p1 = 'cbaebabacd', 'abc'
	s2, p2 = 'abab', 'ab'
	Solution().findAnagrams(s1, p1) == [0, 6]
	Solution().findAnagrams(s2, p2) == [0, 1, 2]
	print "All tests passed!"